File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/util/modal/ModalList.java
package io.github.arcaneplugins.levelledmobs.bukkit.util.modal;

import io.github.arcaneplugins.levelledmobs.bukkit.util.Log;
import java.util.ArrayList;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.ConfigurateException;

public class ModalList<T> extends ModalCollection<T> {

    public ModalList(
        final @NotNull List<T> items,
        final @NotNull ModalCollection.Mode mode
    ) {
        super(items, mode);
    }

    public static ModalList<String> parseModalStringListFromNode(
        final CommentedConfigurationNode node
    ) {
        try {
            if(node.hasChild("in-list")) {
                return new ModalList<>(
                    node.node("in-list").getList(String.class, new ArrayList<>()),
                    Mode.INCLUSIVE
                );
            } else if(node.hasChild("not-in-list")) {
                return new ModalList<>(
                    node.node("not-in-list").getList(String.class, new ArrayList<>()),
                    Mode.EXCLUSIVE
                );
            } else {
                throw new IllegalStateException(
                    "Modal list at path '" + node.path().toString() +
                        "' does not have 'in-list' or 'not-in-list' declared.");
            }
        } catch(ConfigurateException ex) {
            Log.sev("Unable to parse modal list at path: " + node.path().toString(), true);
            throw new RuntimeException(ex);
        }
    }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/util/modal/ModalList.kt
package io.github.arcaneplugins.levelledmobs.bukkit.util.modal

import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.sev
import org.spongepowered.configurate.CommentedConfigurationNode
import org.spongepowered.configurate.ConfigurateException

class ModalList<T>(
    items: MutableList<T>,
    mode: Mode
) : ModalCollection<T>(items, mode) {
    companion object{
        fun parseModalStringListFromNode(
            node: CommentedConfigurationNode
        ): ModalList<String>{
            var result: ModalList<String>? = null
            try {
                if (node.hasChild("in-list")) {
                    result = ModalList(
                        node.node("in-list").getList(String::class.java, mutableListOf()),
                        Mode.INCLUSIVE
                    )
                } else if (node.hasChild("not-in-list")) {
                    result = ModalList(
                        node.node("not-in-list").getList(String::class.java, mutableListOf()),
                        Mode.EXCLUSIVE
                    )
                } else {
                    throw IllegalStateException(
                        "Modal list at path '" + node.path().toString() +
                                "' does not have 'in-list' or 'not-in-list' declared."
                    )
                }
            } catch (ex: ConfigurateException) {
                sev("Unable to parse modal list at path: " + node.path().toString(), true)
                throw RuntimeException(ex)
            }
            return result
        }
    }
}